### PR TITLE
BUG: Fix for duplicate authentication

### PR DESCRIPTION
**Pull request by ant0nsc**
Resolve issues with AML SDK v1/v2: Even in the v2 code path we are creating v1 workspaces, leading to duplicate authentication requests.
To achieve that, a couple of other changes were necessary:
* Deprecate the use of the default datastore, which was read out of an SDK v1 Workspace object even if SDK v2 was chosen
* No longer allowing SDK v2 when mounting datasets for local runs (v2 Datasets can't be mounted at all).

Also added more detailed logging for dataset creation, and a commandline flag to control logging level.

### CODE REVIEW

File: hi-ml-azure/src/health_azure/datasets.py [hunk 1]
```diff
@@ -128,11 +130,14 @@ def _get_or_create_v1_dataset(datastore_name: str, dataset_name: str, workspace:
...
         azureml_dataset = _retrieve_v1_dataset(dataset_name, workspace)
     except UserErrorException:
+        logger.warning(f"Dataset '{dataset_name}' was not found, or is not an AzureML SDK v1 dataset.")
+        logger.info(f"Trying to creating a new dataset '{dataset_name}' from files in folder '{dataset_name}'")
         if datastore_name == "":
             raise ValueError(
                 "When creating a new dataset, a datastore name must be provided. Please specify a datastore name using "
                 "the --datastore flag"
             )
+        logger.info(f"Trying to creating a new dataset '{dataset_name}' in datastore '{datastore_name}'")
```
Comment (fepegar) [inline hi-ml-azure/src/health_azure/datasets.py:140]:
```suggestion
        logger.info(f"Trying to create a new dataset '{dataset_name}' in datastore '{datastore_name}'")
```

**Comment by fepegar** [comment]
Can you please summarise the changes in this PR so it's a bit easier to review?

File: hi-ml/src/health_ml/experiment_config.py [hunk 2]
```diff
@@ -87,10 +95,23 @@ class ExperimentConfig(param.Parameterized):
...
         class_=RunnerMode,
         default=RunnerMode.TRAIN,
         doc=f"The mode to run the experiment in. Can be one of '{RunnerMode.TRAIN}' (training and evaluation on the "
         f"test set), or '{RunnerMode.EVAL_FULL}' for evaluation on the full dataset specified by the "
         "'get_eval_data_module' method of the container.",
     )
+    log_level: Optional[RunnerMode] = param.ClassSelector(
+        class_=LogLevel,
+        default=None,
+        doc=f"The log level to use. Can be one of '{LogLevel.ERROR}', '{LogLevel.WARNING}', '{LogLevel.INFO}', "
```
Comment (kenza-bouzid) [inline hi-ml/src/health_ml/experiment_config.py:108]: replace log levels in docu by a list, note that if you add a loglevel in teh future you'll have to remember updating the docu as well, using a list is more maintainable
```suggestion
        doc=f"The log level to use. Can be one of {list(map(str, LogLevel))} "
```

**Review by kenza-bouzid — APPROVED** [review]
LGTM, minor suggestions and Qs

File: hi-ml-azure/src/health_azure/himl.py [hunk 3]
```diff
@@ -935,8 +938,9 @@ def submit_to_azure_if_needed(  # type: ignore
 
     with append_to_amlignore(amlignore=amlignore_path, lines_to_append=lines_to_append):
         if strictly_aml_v1:
+            assert aml_workspace is not None
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-azure/src/health_azure/himl.py:941]: I've resolved it different by distinguishing between the arguments and the resolved workspace/ml_clients which are never None

File: hi-ml-azure/src/health_azure/himl.py [hunk 4]
```diff
@@ -895,16 +894,23 @@ def submit_to_azure_if_needed(  # type: ignore
...
         logs_folder.mkdir(exist_ok=True)
 
+        any_local_folders_missing = any(dataset.local_folder is None for dataset in cleaned_input_datasets)
+
+        if has_input_datasets and any_local_folders_missing and not strictly_aml_v1:
+            raise ValueError(
+                "AzureML SDK v2 does not support downloading datasets from AzureML for local execution. "
+                "Please switch to AzureML SDK v1 by removing the strictly_aml_v1==False flag, or use "
+                "--strictly_aml_v1 on the commandline, or provide a local folder for each input dataset. "
+                "Note that you will not be able to download datasets that were created via SDK v2."
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-azure/src/health_azure/himl.py:904]: better now?

File: hi-ml-azure/src/health_azure/himl.py [hunk 5]
```diff
@@ -895,16 +894,23 @@ def submit_to_azure_if_needed(  # type: ignore
...
-            strictly_aml_v1,
-            aml_workspace=aml_workspace,
-            ml_client=ml_client,
-            workspace_config_path=workspace_config_path,
+            workspace=aml_workspace,
         )
 
         return AzureRunInfo(
-            input_datasets=mounted_input_datasets,
+            input_datasets=mounted_input_datasets,  # type: ignore
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-azure/src/health_azure/himl.py:913]: Mypy being too strict - argument is type List[Path | None], but I'm giving it List[Path ]

File: hi-ml-azure/src/health_azure/himl.py [hunk 6]
```diff
@@ -882,6 +868,19 @@ def submit_to_azure_if_needed(  # type: ignore
     # is necessary. If not, return to the caller for local execution.
     if submit_to_azureml is None:
         submit_to_azureml = AZUREML_FLAG in sys.argv[1:]
+
+    has_input_datasets = len(cleaned_input_datasets) > 0
+
+    if submit_to_azureml or has_input_datasets:
+        if strictly_aml_v1:
+            aml_workspace = get_workspace(aml_workspace, workspace_config_path)
+            assert aml_workspace is not None
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-azure/src/health_azure/himl.py:877]: resolved - this only came from using the argument which is Workspace | None

File: hi-ml-azure/src/health_azure/himl.py [hunk 7]
```diff
@@ -494,18 +488,6 @@ def submit_run_v2(
...
-                subscription_id=workspace.subscription_id,
-                resource_group=workspace.resource_group,
-                workspace_name=workspace.name,
-            )
-        elif workspace_config_path is not None:
-            ml_client = get_ml_client(workspace_config_path=workspace_config_path)
-        else:
-            raise ValueError("Either workspace or workspace_config_path must be specified to connect to the Workspace")
-
     assert compute_target is not None, "No compute_target has been provided"
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-azure/src/health_azure/himl.py:491]: good observation. Out of scope for this PR, but changing nevertheless. There's probably tons of small such niggles here.

File: hi-ml-azure/src/health_azure/utils.py [hunk 8]
```diff
@@ -1910,6 +1934,7 @@ def get_credential() -> Optional[TokenCredential]:
     tenant_id = get_secret_from_environment(ENV_TENANT_ID, allow_missing=True)
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-azure/src/health_azure/utils.py:1934]: This is for consistency between v1 and v2.

File: hi-ml-azure/testazure/testazure/test_get_ml_client.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+        cred = _get_legitimate_default_credential()
+        assert isinstance(cred, DefaultAzureCredential)
+
+
+@pytest.mark.fast
+def test_get_ml_client_with_existing_client() -> None:
+    """When passing an existing ml_client, it should be returned"""
+    ml_client = "mock_ml_client"
+    result = get_ml_client(ml_client=ml_client)  # type: ignore
+    assert result == ml_client
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-azure/testazure/testazure/test_get_ml_client.py:173]: yeah, it's a bit hacky - that's why the type: ignore is there

File: hi-ml-azure/src/health_azure/himl.py [hunk 10]
```diff
@@ -898,16 +894,24 @@ def submit_to_azure_if_needed(  # type: ignore
         logs_folder = Path.cwd() / LOGS_FOLDER
         logs_folder.mkdir(exist_ok=True)
 
+        any_local_folders_missing = any(dataset.local_folder is None for dataset in cleaned_input_datasets)
+
+        if has_input_datasets and any_local_folders_missing and not strictly_aml_v1:
+            raise ValueError(
+                "AzureML SDK v2 does not support downloading datasets from AzureML for local execution. "
+                "Please switch to AzureML SDK v1 by removing the strictly_aml_v1=False flag, or use "
```
Comment (fepegar) [inline hi-ml-azure/src/health_azure/himl.py:902]:
```suggestion
                "Please switch to AzureML SDK v1 by setting strictly_aml_v1=True, or use "
```

File: hi-ml-azure/src/health_azure/himl.py [hunk 11]
```diff
@@ -885,6 +869,18 @@ def submit_to_azure_if_needed(  # type: ignore
     # is necessary. If not, return to the caller for local execution.
     if submit_to_azureml is None:
         submit_to_azureml = AZUREML_FLAG in sys.argv[1:]
+
+    has_input_datasets = len(cleaned_input_datasets) > 0
+    actual_aml_workspace: Optional[Workspace] = None
```
Comment (fepegar) [inline hi-ml-azure/src/health_azure/himl.py:874]: Why do we need `actual_` for these variables?

File: hi-ml-azure/src/health_azure/utils.py [hunk 12]
```diff
@@ -1927,66 +1952,76 @@ def get_credential() -> Optional[TokenCredential]:
...
+            f"Using MLClient for AzureML workspace {ml_client.workspace_name} as specified in config file"
+            f"{workspace_config_path}"
+        )
+        return ml_client
+
+    logger.info("Trying to load the environment variables that define the workspace.")
+    workspace_name = get_secret_from_environment(ENV_WORKSPACE_NAME, allow_missing=True)
+    subscription_id = get_secret_from_environment(ENV_SUBSCRIPTION_ID, allow_missing=True)
+    resource_group = get_secret_from_environment(ENV_RESOURCE_GROUP, allow_missing=True)
+    if bool(workspace_name) and bool(subscription_id) and bool(resource_group):
```
Comment (fepegar) [inline hi-ml-azure/src/health_azure/utils.py:2006]: Are these `bool`s needed?

File: hi-ml-azure/testazure/testazure/test_get_ml_client.py [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+Tests for health_azure.azure_get_workspace and related functions.
+"""
+import os
+from pathlib import Path
+from unittest.mock import DEFAULT, MagicMock, patch
+
+import pytest
+from azure.core.exceptions import ClientAuthenticationError
+from azure.identity import ClientSecretCredential, DefaultAzureCredential, DeviceCodeCredential
+from azureml.exceptions._azureml_exception import UserErrorException
```
Comment (fepegar) [inline hi-ml-azure/testazure/testazure/test_get_ml_client.py:15]: There are a couple of unused variables in this file.

**Review by fepegar — COMMENTED** [review]
LGTM

**Review by fepegar — APPROVED** [review]

File: hi-ml-azure/src/health_azure/himl.py (hunk 11 above)
Comment (ant0nsc) [reply hi-ml-azure/src/health_azure/himl.py:874]: Based on previous PR comments. I thought I could get the code to require fewer asserts just to satisfy mypy, but in the
end, it got as ugly as it was before. Remove the actual... now.

**Pull request merged**